File: relay/hub/atlas.js
// vim: ts=4:sw=4:expandtab

'use strict';

const DEFAULT_ATLAS_URL = 'https://atlas.forsta.io';

const fetch = require('./fetch');
const storage = require('../storage');
const util = require('../util');

const credStoreKey = 'atlasCredential';
const urlStoreKey = 'atlasUrl';


function atobJWT(str) {
    /* See: https://github.com/yourkarma/JWT/issues/8 */
    return Buffer.from(str.replace(/_/g, '/').replace(/-/g, '+'), 'base64').toString('binary');
}

function decodeJWT(encoded_token) {
    let token;
    try {
        const parts = encoded_token.split('.').map(atobJWT);
        token = {
            header: JSON.parse(parts[0]),
            payload: JSON.parse(parts[1]),
            secret: parts[2]
        };
    } catch(e) {
        throw new Error('Invalid Token');
    }
    if (!token.payload || !token.payload.exp) {
        throw TypeError("Invalid Token");
    }
    if (token.payload.exp * 1000 <= Date.now()) {
        throw Error("Expired Token");
    }
    return token;
}


class AtlasClient {

    constructor({url=DEFAULT_ATLAS_URL, jwt=null}) {
        this.url = url;
        if (jwt) {
            const jwtDict = decodeJWT(jwt);
            this.userId = jwtDict.payload.user_id;
            this.orgId = jwtDict.payload.org_id;
            this.authHeader = `JWT ${jwt}`;
        }
    }

    static async factory() {
        const url = await storage.getState(urlStoreKey);
        const jwt = await storage.getState(credStoreKey);
        return new this({url, jwt});
    }

    static async requestAuthenticationCode(userTag, options) {
        const client = new this(options || {});
        const [user, org] = client.parseTag(userTag);
        await client.fetch(`/v1/login/send/${org}/${user}/`);
        return smsCode => this.authenticateViaCode(userTag, smsCode, options);
    }

    static async authenticateViaCode(userTag, code, options) {
        const client = new this(options || {});
        const [user, org] = client.parseTag(userTag);
        const auth = await client.fetch('/v1/login/authtoken/', {
            method: 'POST',
            json: {authtoken: [org, user, code].join(':')}
        });
        await storage.putState(credStoreKey, auth.token);
        await storage.putState(urlStoreKey, client.url);
        return new this(Object.assign({
            url: client.url,
            jwt: auth.token
        }, options || {}));
    }

    static async authenticateViaToken(userAuthToken, options) {
        const client = new this(options || {});
        const auth = await client.fetch('/v1/login/authtoken/', {
            method: 'POST',
            json: {userauthtoken: userAuthToken}
        });
        await storage.putState(credStoreKey, auth.token);
        await storage.putState(urlStoreKey, client.url);
        return new this(Object.assign({
            url: client.url,
            jwt: auth.token
        }, options || {}));
    }

    parseTag(tag) {
        tag = tag.replace(/^@/, '');
        const index = tag.indexOf(':');
        if (index === -1) {
            return [tag, 'forsta'];
        } else {
            return [tag.substring(0, index), tag.substring(index + 1)];
        }
    }

    async fetch(urn, options) {
        options = options || {};
        options.headers = options.headers || new fetch.Headers();
        if (this.authHeader) {
            options.headers.set('Authorization', this.authHeader);
        }
        const url = [this.url, urn.replace(/^\//, '')].join('/');
        const resp = await fetch(url, options);
        const text = await resp.text();
        let json = undefined;
        if ((resp.headers.get('content-type') || '').startsWith('application/json') && text.trim()) {
            json = JSON.parse(text);
        } 
        if (!resp.ok) {
            const msg = `${urn} (${text})`;
            throw new util.RequestError(msg, resp, resp.status, text, json);
        }
        return json === undefined ? text : json;
    }

    async maintainJWT(forceRefresh, authenticator, onRefresh) {
        /* Manage auth token expiration.  This routine will reschedule itself as needed. */
        let token = decodeJWT(await storage.getState(credStoreKey));
        const refreshDelay = t => (t.payload.exp - (Date.now() / 1000)) / 2;
        if (forceRefresh || refreshDelay(token) < 1) {
            const encodedToken = await storage.getState(credStoreKey);
            const resp = await this.fetch('/v1/api-token-refresh/', {
                method: 'POST',
                json: {token: encodedToken}
            });
            let jwt;
            if (!resp || !resp.token) {
                if (authenticator) {
                    const result = await authenticator();
                    console.info("Reauthenticated user in maintainJWT");
                    jwt = result.jwt;
                } else {
                    throw new TypeError("Unable to reauthenticate in maintainJWT");
                }
            } else {
                jwt = resp.token;
            }
            token = decodeJWT(jwt);
            console.info("Refreshed JWT in maintainJWT");
            await storage.putState(credStoreKey, jwt);
            this.authHeader = `JWT ${jwt}`;
            this.userId = token.payload.user_id;
            if (onRefresh) {
                try {
                    await onRefresh(token);
                } catch(e) {
                    console.error('onRefresh callback error:', e);
                }
            }
        }
        const nextUpdate = refreshDelay(token);
        console.info('maintainJWT will recheck auth token in ' + nextUpdate + ' seconds');
        util.sleep(nextUpdate).then(this.maintainJWT.bind(this, false, authenticator, onRefresh));
    }

    async resolveTags(expression) {
        return (await this.resolveTagsBatch([expression]))[0];
    }

    async resolveTagsBatch(expressions) {
        if (!expressions.length) {
            return [];
        }
        const resp = await this.fetch('/v1/tagmath/', {
            method: 'POST',
            json: {expressions}
        });
        /* Enhance the warnings a bit. */
        for (let i = 0; i < resp.results.length; i++) {
            const res = resp.results[i];
            const expr = expressions[i];
            for (const w of res.warnings) {
                w.context = expr.substr(w.position, w.length);
            }
        }
        return resp.results;
    }

    sanitizeTags(expression) {
        /* Clean up tags a bit. Add @ where needed.
         * NOTE: This does not currently support universal format! */
        const tagSplitRe = /([\s()^&+-]+)/;
        const tags = [];
        for (let tag of expression.trim().split(tagSplitRe)) {
            if (!tag) {
                continue;
            } else if (tag.match(/^[a-zA-Z]/)) {
                tag = '@' + tag;
            }
            tags.push(tag);
        }
        return tags.join(' ');
    }

    async getUsers(userIds, onlyDir) {
        const missing = new Set(userIds);
        const users = [];
        if (!onlyDir) {
            const resp = await this.fetch('/v1/user/?id_in=' + userIds.join());
            for (const user of resp.results) {
                users.push(user);
                missing.delete(user.id);
            }
        }
        if (missing.size) {
            const resp = await this.fetch('/v1/directory/user/?id_in=' +
                                          Array.from(missing).join());
            for (const user of resp.results) {
                users.push(user);
            }
        }
        return users;
    }

    async getDevices() {
        try {
            return (await this.fetch('/v1/provision/account')).devices;
        } catch(e) {
            if (e instanceof util.RequestError && e.code === 404) {
                return [];
            } else {
                throw e;
            }
        }
    }
}

module.exports = AtlasClient;

File: relay/hub/atlas.py
"""
Atlas API client
"""

import aiohttp
import asyncio
import base64
import json
import logging
import re
import time
from . import http
from .. import storage

store = storage.getStore()
logger = logging.getLogger(__name__)

DEFAULT_ATLAS_URL = 'https://atlas.forsta.io'
cred_store_key = 'atlasCredential'
url_store_key = 'atlasUrl'


def jwt_b64decode(value):
    padding = 4 - (len(value) % 4)
    return base64.urlsafe_b64decode(value + ("=" * padding))


def decode_jwt(encoded_token):
    parts = [jwt_b64decode(x) for x in encoded_token.split('.')]
    token = {
        "header": json.loads(parts[0]),
        "payload": json.loads(parts[1]),
        "secret": parts[2]
    }
    if not token['payload'] or not token['payload']['exp']:
        raise TypeError("Invalid Token")
    if token['payload']['exp'] <= time.time():
        raise ValueError("Expired Token")
    return token


class AtlasClient(http.HttpClient):

    def __init__(self, url=None, jwt=None):
        self.setJWT(jwt)
        super().__init__(url=url or DEFAULT_ATLAS_URL)

    def setJWT(self, jwt):
        if jwt is not None:
            jwtDict = decode_jwt(jwt)
            self.userId = jwtDict['payload']['user_id']
            self.orgId = jwtDict['payload']['org_id']
            self._authHeader = 'JWT ' + jwt
        else:
            self.userId = None
            self.orgId = None
            self._authHeader = None

    def authHeader(self):
        return self._authHeader

    @classmethod
    def factory(cls):
        url = store.getState(url_store_key)
        jwt = store.getState(cred_store_key)
        return cls(url=url, jwt=jwt)

    @classmethod
    async def requestAuthentication(cls, userTag, **options):
        client = cls(**options)
        user, org = client.parseTag(userTag)
        async with client.fetchRequest(f'/v1/login/send/{org}/{user}/') as resp:
            data = await resp.json()
            if resp.status == 409:
                if 'totp auth required' in data['non_field_errors']:
                    auth_type = 'totp'

                    async def validator(password, otp):
                        return await cls.authenticateViaPasswordOtp(userTag,
                            password, otp, **options)
                else:
                    auth_type = 'password'

                    async def validator(password):
                        return await cls.authenticateViaPassword(userTag, password, **options)
            else:
                auth_type = 'sms'

                async def validator(code):
                    return await cls.authenticateViaCode(userTag, code, **options)
            return auth_type, validator

    @classmethod
    async def requestAuthenticationCode(cls, userTag, **options):
        client = cls(**options)
        user, org = client.parseTag(userTag)
        await client.fetch(f'/v1/login/send/{org}/{user}/')
        return lambda code: cls.authenticateViaCode(userTag, code, **options)

    @classmethod
    async def authenticateViaCode(cls, userTag, code, **options):
        client = cls(**options)
        user, org = client.parseTag(userTag)
        await client.authenticate(authtoken=f'{org}:{user}:{code}')
        return client

    @classmethod
    async def authenticateViaToken(cls, userauthtoken, **options):
        client = cls(**options)
        await client.authenticate(userauthtoken=userauthtoken)
        return client

    @classmethod
    async def authenticateViaPassword(cls, fq_tag, password, **options):
        client = cls(**options)
        await client.authenticate(fq_tag=fq_tag, password=password)
        return client

    @classmethod
    async def authenticateViaPasswordOtp(cls, fq_tag, password, otp, **options):
        client = cls(**options)
        await client.authenticate(fq_tag=fq_tag, password=password, otp=otp)
        return client

    async def authenticate(self, **creds):
        """ Creds should be an object of these supported forms..
            1. Password auth:
                 fq_tag: "@foo:bar",
                 password: "secret"
            1.5 Password+TOTP auth:
                 fq_tag: "@foo:bar",
                 password: "secret"
                 otp: "code"
            2. SMS auth:
                 authtoken: "123456",
            3. Token auth:
                 userauthtoken: "APITOKEN",
        """
        auth = await self.fetch('/v1/login/', method='POST', json=creds)
        self.setJWT(auth['token'])
        store.putState(cred_store_key, auth['token'])
        store.putState(url_store_key, self.url)

    def parseTag(self, tag):
        user, *org = tag.lstrip('@').split(':', 1)
        if not org:
            org = ['forsta']
        return user, org[0]

    async def maintainJWT(self, forceRefresh=False, authenticator=None,
                          onRefresh=None):
        """ Manage auth token expiration.  This routine will reschedule itself
        as needed. """
        token = decode_jwt(store.getState(cred_store_key))
        refresh_delay = lambda t: (t['payload']['exp'] - time.time()) / 2
        if forceRefresh or refresh_delay(token) < 1:
            encoded_token = store.getState(cred_store_key)
            resp = await self.fetch('/v1/api-token-refresh/', method="POST",
                                    json={"token": encoded_token})
            if not resp or not resp['token']:
                if authenticator:
                    result = await authenticator()
                    logger.info("Reauthenticated user in maintainJWT")
                    jwt = result.jwt
                else:
                    raise TypeError("Unable to reauthenticate in maintainJWT")
            else:
                jwt = resp['token']
            token = decode_jwt(jwt)
            logger.info("Refreshed JWT in maintainJWT")
            store.putState(cred_store_key, jwt)
            self.setJWT(jwt)
            if onRefresh:
                try:
                    await onRefresh(token)
                except Exception as e:
                    logger.exception('on_refresh callback error')
        next_update = refresh_delay(token)
        logger.info('maintainJWT will recheck auth token in %f seconds' %
                    next_update)
        await asyncio.sleep(next_update)
        await self.maintainJWT(False, authenticator, onRefresh)

    async def resolveTags(self, expression):
        return (await self.resolveTagsBatch([expression]))[0]

    async def resolveTagsBatch(self, expressions):
        if not expressions:
            return []
        resp = await self.fetch('/v1/tagmath/', method='POST',
                                json={"expressions": expressions})
        # Enhance the warnings a bit.
        for res, expr in zip(resp['results'], expressions):
            for w in res['warnings']:
                w['context'] = expr[w['position']:w['position']+w['length']]
        return resp['results']

    def sanitizeTags(self, expression):
        """ Clean up tags a bit. Add @ where needed.
        NOTE: This does not currently support universal format! """
        sep_re = re.compile(r'([\s()^&+-]+)')
        tags = []
        for tag in filter(None, sep_re.split(expression.strip())):
            if re.compile(r'^[a-zA-Z]').match(tag):
                tag = '@' + tag
            tags.append(tag)
        return ' '.join(tags)

    async def getUsers(self, userIds, onlyDir=False):
        missing = set(userIds)
        users = []
        if not onlyDir:
            resp = await self.fetch('/v1/user/?id_in=' + ','.join(userIds))
            for user in resp['results']:
                users.append(user)
                missing.remove(user.id)
        if missing:
            resp = await self.fetch('/v1/directory/user/?id_in=' +
                                    ','.join(missing))
            for user in resp['results']:
                users.append(user)
        return users

    async def getDevices(self):
        try:
            return (await self.fetch('/v1/provision/account'))['devices']
        except aiohttp.ClientResponseError as e:
            if e.status == 404:
                return []
            else:
                raise e
